Public/Src/Cache/ContentStore/Distributed/NuCache/ContentLocationDatabase.cs

```diff
@@ -658,10 +658,17 @@ private void LogEntryDeletion(OperationContext context, ShortHash hash, ContentL
         /// <param name="contentHashListWithDeterminism">
         ///     The value, and associated determinism guarantee, to store.
         /// </param>
+        /// <param name="contentSession">
+        ///     The content session over which this operation is happening
+        /// </param>
+        /// <param name="cts">
+        ///     A token that can signal this call should return as soon as possible.
+        /// </param>
         /// <returns>
         ///     Result providing the call's completion status.
         /// </returns>
-        public abstract AddOrGetContentHashListResult AddOrGetContentHashList(OperationContext context, StrongFingerprint strongFingerprint, ContentHashListWithDeterminism contentHashListWithDeterminism);
+        public abstract Task<AddOrGetContentHashListResult> AddOrGetContentHashListAsync(OperationContext context, StrongFingerprint strongFingerprint, ContentHashListWithDeterminism contentHashListWithDeterminism, IContentSession contentSession,

```

Comment:
Need to discuss this with Lance, but it looks like we're mixing the layers here: I think that the database is relatively low level concept, but now its aware of IContentSession.

I'm not sure for now how easy/doable/wise to avoid this coupling though.

Reply:
Indeed. Database should not be aware of IContentSession. Any logic requiring this should be moved up into the wrapping IMemoizationSession/Store implementation.

---
In reply to: [305161161](https://github.com/microsoft/BuildXL/pull/621#discussion_r305161161) [](ancestors = 305161161)